### ISSUE TITLE
Diagnostic for argument constraint with non-nullable type instead of nullable type

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1060]
Split from #1005. Inspired by #660.

Assuming you have this interface:
```c#
public interface IFoo
{
    int Bar(int? x);
}
```
And a call configuration like this:
```c#
A.CallTo(() => foo.Bar(A<int>._)).Returns(42);
```
It's easy to miss the fact that the constraint won't match when the argument is `null` (it used to throw a NRE, which at least indicated something was wrong, but now it silently rejects the null argument). It's especially confusing because when you write `A<int>._`, you usually mean "match anything". The analyzer could detect this and emit a warning.

(I wish we could omit the type and just say `A.CallTo(() => foo.Bar(A._)`, which is more concise and easier to get right. Java's type inference is a little better in that regard. This syntax could also be achieved if we had a ["bottom type"](https://en.wikipedia.org/wiki/Bottom_type) in C#.)

**Comment by thomaslevesque** [comment]
The only problem is that it could be intentional, if you specifically don't want to include nulls... this would make the warning a false positive. Maybe we should provide a better way to specify "not null":

    A.CallTo(() => foo.Bar(A<int?>.That.IsNotNull())).Returns(42);

**Comment by thomaslevesque** [comment]
> (interestingly, we have a IsNull constraint, which isn't very useful since we can just say null, but no IsNotNull...)

Of course, I forgot about `Not`... we can do `A<T>.That.Not.IsNull()`. But it doesn't read very well...

**Comment by thomaslevesque** [comment]
The warning could have a codefix to change `A<int>._` to either `A<int?>._` or `A<int?>.That.Not.IsNull()`.

**Comment by blairconrad** [comment]
> > (interestingly, we have a IsNull constraint, which isn't very useful since we can just say null, but no IsNotNull...)

> Of course, I forgot about `Not`... we can do `A<T>.That.Not.IsNull()`. But it doesn't read very well...

I'm usually against adding extra mechanisms to accomplish things we already support, but "not null" is such a fundamental operation that I think the overload would be worthwhile.
Shall we just do it regardless?

**Comment by thomaslevesque** [comment]
> Shall we just do it in advance of this analyzer issue?

👍

**Comment by thomaslevesque** [comment]
I have a working analyzer on [this branch](https://github.com/thomaslevesque/FakeItEasy/commits/nullable-constraint-analyzer). I'd like to add a code fix before I submit a PR, but I'm not sure what to do about constraints that use `A<T>.That`... The fix I suggested above should work fine for `_` and `Ignored`, but `That` can introduce an arbitrary constraint, so we can't easily suggest a fix. In fact, I wonder if the analyzer should even warn about that case...

**Comment by blairconrad** [comment]
False positives are the worst of all situations, so if the `That` analysis might overreport warnings, I would leave it out. It's easy to add new cases to the analyzer later.

A fix for `Ignored` (and `_`) sounds grand!

**Comment by thomaslevesque** [comment]
> False positives are the worst of all situations, so if the That analysis might overreport warnings, I would leave it out. It's easy to add new cases to the analyzer later.
> 
> A fix for Ignored (and _) sounds grand!

The thing is, this analyzer is bound to cause false positives, because you can intentionally write `A<int>._` where a `int?` is expected. This is why I want to provide a code-fix to easily change it to something that is equivalent and more obviously correct. But since I can't do that for `A<T>.That...`, the user would have to suppress the warning if it was intentional.

So I think I'll just restrict the warning to `Ignored` and `_`.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.0.0).
